kernel/nano_core/src/bios.rs

```diff
@@ -1,25 +1,25 @@
 //! The main entry point into Rust code from a legacy BIOS (multiboot2) bootloader.
 
-use crate::{nano_core, shutdown, try_exit};
+use crate::{nano_core, shutdown};
 use boot_info::BootInformation;
 use memory::VirtualAddress;
 
 #[no_mangle]
 pub extern "C" fn rust_entry(boot_info_vaddr: usize, double_fault_stack_top: usize) {
-    if VirtualAddress::new(boot_info_vaddr).is_none() {
-        shutdown(format_args!("BUG: multiboot2 info virtual address is invalid"));
-    }
-    let boot_info = match unsafe { multiboot2::load(boot_info_vaddr) } {
-        Ok(i) => i,
-        Err(e) => shutdown(format_args!("BUG: failed to load multiboot 2 info: {e:?}")),
-    };
-    let kernel_stack_start = try_exit!(
-        VirtualAddress::new(double_fault_stack_top - try_exit!(boot_info.stack_size()))
-            .ok_or("BUG: kernel_stack_start virtual address is invalid")
+    setup(boot_info_vaddr, double_fault_stack_top).unwrap_or_else(|e|
+        shutdown(format_args!("{e}"))
     );
-    let double_fault_stack_top = try_exit!(
-        VirtualAddress::new(double_fault_stack_top)
-            .ok_or("BUG: double_fault_stack_top virtual address is invalid")
-    );
-    try_exit!(nano_core(boot_info, double_fault_stack_top, kernel_stack_start));
+}
+
+fn setup(boot_info_vaddr: usize, double_fault_stack_top: usize) -> Result<(), &'static str> {
+    VirtualAddress::new(boot_info_vaddr)
+        .ok_or("BUG: multiboot2 info virtual address is invalid")?;
+    let boot_info = unsafe { multiboot2::load(boot_info_vaddr) }.ok()
+        .ok_or("BUG: failed to load multiboot 2 info")?;
+    let kernel_stack_start = VirtualAddress::new(double_fault_stack_top - boot_info.stack_size()?)
+        .ok_or("BUG: kernel_stack_start virtual address is invalid")?;
+    let double_fault_stack_top = VirtualAddress::new(double_fault_stack_top)
+        .ok_or("BUG: double_fault_stack_top virtual address is invalid")?;
+
+    nano_core(boot_info, double_fault_stack_top, kernel_stack_start)
```

Comment:
Hmm, i guess my question is... why? This isn't really any better than what we had, and you've discarded info from the `multiboot2::load` error condition.

Reply:
I tried to get that error condition by propagating format_args but it looked worse, then I thought "when is this going to happen" and decided that if it was to happen, it would still be easy enough to debug without the info. If the multiboot Error  just implemented Error or to_string or sth and or we also had error types here, it would be easier.

Besides that, I think this is easier to read and removes one more macro. The entry point just handles control flow setup -> shutdown and preferrably this would be the only shutdown-point in theseus.

Reply:
fair enough